CMakeLists.txt

```diff
@@ -1,220 +1,177 @@
 
-SET_PROPERTY(GLOBAL PROPERTY USE_FOLDERS ON)
-if(POLICY CMP0058)
-  CMAKE_POLICY(SET CMP0058 NEW)
-endif(POLICY CMP0058)
+if(APPLE)
+  # Brew does not put the glbinding cmake config files where it can be found
+  # TODO(umar) check if other systems have a similar problem
+  set(CMAKE_PREFIX_PATH "/usr/local/opt/glbinding;${CMAKE_PREFIX_PATH}")
+  # TODO(umar) Remove rpath to third_part lib
+  set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/${AF_INSTALL_LIB_DIR};${ArrayFire_BINARY_DIR}/third_party/forge/lib")
 
-SET(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_CURRENT_SOURCE_DIR}/CMakeModules")
-INCLUDE(UploadCoveralls)
-INCLUDE(AFInstallDirs)
+endif()
 
-OPTION(BUILD_TEST "Build Tests" ON)
-OPTION(BUILD_EXAMPLES "Build Examples" ON)
+include(build_cl2hpp)
+
+include(CMakeDependentOption)
+include(ConditionalDirectory)
+include(DependencyCheck)
+include(AFInstallDirs)
+include(Version)
+
+find_package(CUDA 7.0)
+find_package(OpenCL 1.2)
+find_package(OpenGL)
+find_package(OpenMP)
+find_package(FreeImage)
+find_package(Threads)
+find_package(FFTW)
+find_package(CBLAS)
+find_package(LAPACKE)
+find_package(Doxygen)
+
+# Graphics dependencies
+find_package(glbinding QUIET)
+find_package(Boost)
+
+option(BUILD_CPU      "Build ArrayFire with a CPU backend"        ON)
+option(BUILD_CUDA     "Build ArrayFire with a CUDA backend"       ${CUDA_FOUND})
+option(BUILD_OPENCL   "Build ArrayFire with a OpenCL backend"     ${OpenCL_FOUND})
+option(BUILD_UNIFIED  "Build Backend-Independent ArrayFire API"   ON)
+
+option(BUILD_GRAPHICS "Build ArrayFire with Forge Graphics"      $<AND:${OPENGL_FOUND},${Forge_FOUND},${glbinding_FOUND}>)
+option(BUILD_DOCS     "Create ArrayFire Documentation"           ${DOXYGEN_FOUND})
+option(BUILD_NONFREE  "Build ArrayFire nonfree algorithms"       OFF)
+
+option(BUILD_EXAMPLES "Build Examples"                           ON)
+option(BUILD_TEST     "Build Tests"                              ON)
+cmake_dependent_option(USE_RELATIVE_TEST_DIR "Use relative paths for the test data directory(For continious integration(CI) purposes only)" OFF
+  "BUILD_TEST" OFF)
+
+cmake_dependent_option(USE_SYSTEM_FORGE "Use system Forge" OFF
+                    "BUILD_GRAPHICS" OFF)
+
+cmake_dependent_option(WITH_IMAGEIO "Build ArrayFire with Image IO support" ${FreeImage_FOUND}
+                      "FreeImage_FOUND" OFF)
+
+cmake_dependent_option(BUILD_FRAMEWORK "Build an ArrayFire framework for Apple platforms.(Experimental)" OFF
+                      "APPLE" OFF)
+mark_as_advanced(
+  BUILD_FRAMEWORK USE_SYSTEM_FORGE )
+
+option(USE_FREEIMAGE_STATIC "Use Static FreeImage Lib" OFF)
+
+set(USE_CPUID ON CACHE BOOL "Build with CPUID integration")
+mark_as_advanced(USE_CPUID)
+
+# TODO(umar): Add definitions should not be used. Instead use
+# target_include_directories to add definitions.
+if(WIN32)
+  add_definitions(-DOS_WIN -DWIN32_LEAN_AND_MEAN)
+elseif(APPLE)
+  add_definitions(-DOS_MAC)
+elseif(UNIX)
+  add_definitions(-DOS_LNX)
+endif()
```

Comment:
These definitions will be included no matter what target we are compiling. Instead of adding those statements everywhere, I tried to put it in one place.